File: mas.ts
/** Model Address Standard (MAS) parameters. */
export interface MASParams {
  /** Model identifier (required). */
  m: string;
  /** API key (optional). */
  k?: string;
}

/**
 * Extract MAS parameters (`m`, `k`) from an HTTP/HTTPS URI.
 *
 * @param uri - A valid MAS address
 * @returns An object with `m` and optionally `k`
 * @throws If the fragment is missing, `m` is absent, or `m` is empty
 *
 * @example
 * decode("https://api.example.com#m=gpt-4o&k=sk-xyz")
 * // => { m: "gpt-4o", k: "sk-xyz" }
 */
export function decode(uri: string): MASParams {
  const url = new URL(uri);
  const fragment = url.hash.slice(1);
  if (!fragment) {
    throw new Error("MAS: fragment is required");
  }

  const params: Record<string, string> = {};
  for (const part of fragment.split("&")) {
    const idx = part.indexOf("=");
    const key = decodeURIComponent(idx >= 0 ? part.slice(0, idx) : part);
    const value = idx >= 0 ? decodeURIComponent(part.slice(idx + 1)) : "";

    if (key === "m") {
      if (!value) {
        throw new Error("MAS: m must not be empty");
      }
      params["m"] = value;
    } else if (key === "k" && value) {
      params["k"] = value;
    }
  }

  if (!("m" in params)) {
    throw new Error("MAS: m is required");
  }

  return params as unknown as MASParams;
}

/**
 * Build a MAS fragment string from `MASParams`.
 *
 * @param obj - Object with a non-empty `m` and optionally `k`
 * @returns URI fragment of the form `#m=...&k=...`
 * @throws If `m` is missing or empty
 *
 * @example
 * encode({ m: "gpt-4o" })
 * // => "#m=gpt-4o"
 */
export function encode(obj: MASParams): string {
  if (!obj.m) {
    throw new Error("MAS: m is required and must be non-empty");
  }

  const parts = [`m=${encodeURIComponent(obj.m)}`];
  if (obj.k) {
    parts.push(`k=${encodeURIComponent(obj.k)}`);
  }

  return "#" + parts.join("&");
}

File: mas.rb
# Model Address Standard (MAS) — decode/encode MAS URIs.
#
# Usage:
#   MAS.decode("https://api.example.com#m=gpt-4o&k=sk-xyz")
#   # => {"m" => "gpt-4o", "k" => "sk-xyz"}
#   MAS.encode({"m" => "gpt-4o"})
#   # => "#m=gpt-4o"

require "uri"

module MAS
  # Extract MAS parameters (+m+, +k+) from an HTTP/HTTPS URI.
  #
  # @param uri [String] a valid MAS address (any HTTP(S) URI with a
  #   fragment containing at least an +m+ parameter)
  # @return [Hash] hash with key "m" (the model identifier) and
  #   optionally "k" (the API key)
  # @raise [RuntimeError] if the fragment is missing, +m+ is absent,
  #   or +m+ is empty
  def self.decode(uri)
    parsed = URI.parse(uri)
    fragment = parsed.fragment
    raise "MAS: fragment is required" unless fragment

    params = {}
    fragment.split("&").each do |part|
      key, value = part.split("=", 2)
      key = URI.decode_www_form_component(key.to_s)
      value = URI.decode_www_form_component(value.to_s)

      if key == "m"
        raise "MAS: m must not be empty" if value.nil? || value.empty?
        params["m"] = value
      elsif key == "k" && value && !value.empty?
        params["k"] = value
      end
    end

    raise "MAS: m is required" unless params.key?("m")
    params
  end

  # Build a MAS fragment string from a hash with +m+ and optional +k+.
  #
  # @param obj [Hash] hash with a non-empty "m" key and optionally a "k" key
  # @return [String] URI fragment of the form +#m=...&k=...+
  # @raise [RuntimeError] if +m+ is missing or empty
  def self.encode(obj)
    m = obj["m"]
    raise "MAS: m is required and must be non-empty" if m.nil? || m.empty?

    parts = ["m=#{URI.encode_www_form_component(m)}"]
    k = obj["k"]
    parts << "k=#{URI.encode_www_form_component(k)}" if k && !k.empty?

    "##{parts.join("&")}"
  end
end
